### PR TITLE
TASK-5318 - Implement custom annotator to allow clients private files

### PR DESCRIPTION
**Pull request by jtarraga**
Implement a custom extension framework to annotate variants. This requirement implements the COSMIC annotator extension.
[TASK-5318](https://app.clickup.com/t/86937h42v)

### CODE REVIEW

File: opencga-storage/opencga-storage-core/src/main/java/org/opencb/opencga/storage/core/variant/annotation/annotators/extensions/cosmic/CosmicVariantAnnotatorExtensionTask.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+    @Override
+    public List<VariantAnnotation> apply(List<VariantAnnotation> list) throws Exception {
+        for (VariantAnnotation variantAnnotation : list) {
+            Variant variant;
+            try {
+                variant = new Variant(variantAnnotation.getChromosome(), variantAnnotation.getStart(), variantAnnotation.getReference(),
+                        variantAnnotation.getAlternate());
+            } catch (Exception e) {
+                logger.warn("Skipping variant: " + e.getMessage());
```
Comment (j-coll) [inline opencga-storage/opencga-storage-core/src/main/java/org/opencb/opencga/storage/core/variant/annotation/annotators/extensions/cosmic/CosmicVariantAnnotatorExtensionTask.java:120]: Why is this? Which exceptions are you expecting here?

File: opencga-storage/opencga-storage-core/src/main/java/org/opencb/opencga/storage/core/variant/annotation/annotators/extensions/cosmic/CosmicVariantAnnotatorExtensionTask.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        try {
+            // a factory method that returns a RocksDB instance
+            if (indexingNeeded) {
+                rdb = RocksDB.open(dbOption, dbLocation.toAbsolutePath().toString());
+            } else {
+                rdb = RocksDB.openReadOnly(dbOption, dbLocation.toAbsolutePath().toString());
+            }
+        } catch (RocksDBException e) {
+            // Do some error handling
+            throw new ToolException("", e);
```
Comment (j-coll) [inline opencga-storage/opencga-storage-core/src/main/java/org/opencb/opencga/storage/core/variant/annotation/annotators/extensions/cosmic/CosmicVariantAnnotatorExtensionTask.java:170]: No need to wrap the exception. The calling method can throw "Exception", and you're not improving the error message

File: opencga-storage/opencga-storage-core/src/test/java/org/opencb/opencga/storage/core/variant/annotation/VariantAnnotationManagerTest.java [hunk 3]
```diff
@@ -237,6 +242,72 @@ public void testMultiAnnotations() throws Exception {
...
+        runDefaultETL(annotatorExtensionInputUri, variantStorageEngine, newStudyMetadata(),
+                new ObjectMap(VariantStorageOptions.ANNOTATE.key(), false));
+
+        variantStorageEngine.getOptions()
+                .append(VariantStorageOptions.ANNOTATOR_CLASS.key(), DummyTestAnnotator.class.getName())
+                .append(VariantStorageOptions.ANNOTATOR.key(), VariantAnnotatorFactory.AnnotationEngine.OTHER)
+                .append(VariantStorageOptions.ANNOTATOR_EXTENSION_LIST.key(), CosmicVariantAnnotatorExtensionTask.ID)
+                .append(VariantStorageOptions.ASSEMBLY.key(), "GRCh38")
+                .append(VariantStorageOptions.ANNOTATOR_EXTENSION_COSMIC_VERSION.key(), "v95")
+                .append(VariantStorageOptions.ANNOTATOR_EXTENSION_COSMIC_FILE.key(), cosmicFile);
```
Comment (j-coll) [inline opencga-storage/opencga-storage-core/src/test/java/org/opencb/opencga/storage/core/variant/annotation/VariantAnnotationManagerTest.java:261]: Could you please add tests with a wrong cosmic file (missing or invalid format) and/or missing version?

File: opencga-storage/opencga-storage-core/src/test/java/org/opencb/opencga/storage/core/variant/annotation/annotators/extensions/CosmicVariantAnnotatorExtensionTaskTest.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class CosmicVariantAnnotatorExtensionTaskTest {
```
Comment (j-coll) [inline opencga-storage/opencga-storage-core/src/test/java/org/opencb/opencga/storage/core/variant/annotation/annotators/extensions/CosmicVariantAnnotatorExtensionTaskTest.java:25]: Missing category. This looks like a short one, right?

File: opencga-storage/opencga-storage-core/src/main/java/org/opencb/opencga/storage/core/variant/annotation/annotators/extensions/cosmic/CosmicVariantAnnotatorExtensionTask.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    private String assembly;
+
+    private ObjectReader objectReader;
+
+    private RocksDB rdb = null;
+    private Options dbOption = null;
+    private Path dbLocation = null;
+
+    public static final String COSMIC_ANNOTATOR_INDEX_NAME = "cosmicAnnotatorIndex";
+    public static final String COSMIC_VERSION_FILENAME  = "cosmicVersion.json";
```
Comment (j-coll) [inline opencga-storage/opencga-storage-core/src/main/java/org/opencb/opencga/storage/core/variant/annotation/annotators/extensions/cosmic/CosmicVariantAnnotatorExtensionTask.java:45]: This constant is not being used

File: opencga-storage/opencga-storage-core/src/main/java/org/opencb/opencga/storage/core/variant/annotation/annotators/extensions/cosmic/CosmicVariantAnnotatorExtensionTask.java (hunk 1 above)
Comment (jtarraga) [reply opencga-storage/opencga-storage-core/src/main/java/org/opencb/opencga/storage/core/variant/annotation/annotators/extensions/cosmic/CosmicVariantAnnotatorExtensionTask.java:119]: I got `Unknown end or length of the variant '1:1564802-null::<DEL>', type 'DELETION'`; but I think I should use the variant constructor with the parameter `end`, i.e.:

`            Variant variant = new Variant(variantAnnotation.getChromosome(), variantAnnotation.getStart(), variantAnnotation.getEnd(),
                    variantAnnotation.getReference(), variantAnnotation.getAlternate());
`

Comment (j-coll) [reply opencga-storage/opencga-storage-core/src/main/java/org/opencb/opencga/storage/core/variant/annotation/annotators/extensions/cosmic/CosmicVariantAnnotatorExtensionTask.java]: Indeed! You should use that constructor instead, and remove that try-catch

**Review by j-coll — DISMISSED** [review]

**Review by j-coll — APPROVED** [review]

**Review by j-coll — APPROVED** [review]

**Pull request merged**